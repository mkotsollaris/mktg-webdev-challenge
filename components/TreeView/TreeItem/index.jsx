import { useContext } from 'react'
import ArrowIcon from '../../../public/img/arrow.svg'
import style from './style.module.css'
import componentsStyle from '../../style.module.css'
import AppContext from '../../../context/AppContext'

const ExpandIconElement = (
  <img alt="expand" className={style.arrow} src={ArrowIcon} />
)
const CollapseIconElement = (
  <img
    alt="collapse"
    className={`${style['rotate-90']} ${style.arrow}`}
    src={ArrowIcon}
  />
)

const TreeItem = ({ label, children, level }) => {
  const {
    filteredDepartments,
    computeFilteredDepartments,
    expanded,
    setExpanded,
  } = useContext(AppContext)

  const isExpanded = () => {
    return expanded?.includes(label)
  }

  const onClick = () => {
    computeFilteredDepartments(label)

    if (isExpanded()) {
      setExpanded(expanded?.filter((e) => e !== label))
    } else {
      setExpanded([...expanded, label])
    }
  }

  const Icon =
    !children && level === 1 ? (
      <div />
    ) : !children && level > 1 ? null : isExpanded() ? (
      <div>{CollapseIconElement}</div>
    ) : (
      <div>{ExpandIconElement}</div>
    )

  const classNames = filteredDepartments?.includes(label)
    ? `${style.treeItem} ${style.selected}`
    : style.treeItem

  const element = (
    <div className={`${style.treeItem} ${componentsStyle['non-selectable']}`}>
      <button
        tabIndex={0}
        onClick={onClick}
        className={`${style.expandableTreeItem} ${classNames} `}
      >
        {Icon}
        {label}
      </button>
      {isExpanded() ? (
        <ul className={style.ul}>
          <div className={style.children}>{children}</div>
        </ul>
      ) : null}
    </div>
  )
  return element
}

export default TreeItem
