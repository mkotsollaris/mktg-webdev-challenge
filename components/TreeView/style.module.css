.treeView {
  display: grid;
  width: 15rem;
  grid-row-gap: 0.5rem;
}

.secondary-label {
  color: gray;
  font-size: small;
}

.clickable {
  cursor: pointer;
}

.margin-left-sm {
  margin-left: 0.5rem;
}
