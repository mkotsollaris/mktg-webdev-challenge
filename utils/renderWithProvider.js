import { render } from '@testing-library/react'
import AppContext from '../context/AppContext'

function renderWithProvider(props, Component) {
  return render(
    <AppContext.Provider value={props}>
      <Component />
    </AppContext.Provider>
  )
}

export default renderWithProvider
