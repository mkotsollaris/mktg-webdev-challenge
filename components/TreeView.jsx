import { useContext } from "react";
import AppContext from "../context/AppContext";
import TreeItem from "./TreeItem";

const TreeView = () => {

    const { treeNodes } = useContext(AppContext);

    const renderTreeNode = (node, intend = 0) => {

        const hasChildren = node && node.children && node.children.length > 0;

        if (hasChildren) {
            const innerEl = node.children.map(child => renderTreeNode(child, intend + 1))
            return <TreeItem key={node.attributes.id} intend={intend} label={node.value}>
                {innerEl}
            </TreeItem>
        } else {
            return <TreeItem key={node.attributes.id} intend={intend} label={node.value} />
        }
    }

    const content = treeNodes?.map(node => renderTreeNode(node))

    return <>
        {content}
    </>
}

export default TreeView;