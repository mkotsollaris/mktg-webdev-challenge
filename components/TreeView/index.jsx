import { useContext } from 'react'
import AppContext from '../../context/AppContext'
import TreeItem from './TreeItem'
import componentStyle from '../style.module.css'

import style from './style.module.css'
import ClearFilter from '../ClearFilter'

const TreeView = () => {
  const { treeNodes, filteredDepartments } = useContext(AppContext)

  const renderTreeNode = (node) => {
    const hasChildren = node && node.children && node.children.length > 0

    if (hasChildren) {
      const innerEl = node.children.map((child) => renderTreeNode(child))
      return (
        <TreeItem key={node.attributes.id} label={node.value} node={node}>
          {innerEl}
        </TreeItem>
      )
    }
    return <TreeItem key={node.attributes.id} label={node.value} />
  }

  const content = treeNodes?.map((node) => renderTreeNode(node))

  return (
    <div className={style.treeView}>
      <span className={`${componentStyle.name}`}>Filter by department</span>
      {filteredDepartments.length === 0 ? null : <ClearFilter />}
      {content}
    </div>
  )
}

export default TreeView
