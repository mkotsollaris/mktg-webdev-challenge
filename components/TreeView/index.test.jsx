import TreeView from '.'
import renderWithProvider from '../../utils/renderWithProvider'

it('renders', () => {
  const { asFragment } = renderWithProvider({}, TreeView)
  expect(asFragment()).toMatchSnapshot()
})
