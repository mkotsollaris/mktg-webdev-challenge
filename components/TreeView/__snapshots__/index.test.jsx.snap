// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div>
    <span
      class="undefined"
    >
      Filter by department
    </span>
    <button
      class="undefined undefined undefined"
    >
      clear filters
    </button>
  </div>
</DocumentFragment>
`;
